### PR TITLE
Refactor Timestamp class to be safer. Add methods to minus a timestam…

### PR DESCRIPTION
**Pull request by bogdandrutu**
…p and add duration.

### CODE REVIEW

File: api/src/main/java/io/opencensus/common/Timestamp.java [hunk 1]
```diff
@@ -45,101 +46,116 @@ private Timestamp(long seconds, int nanos) {
...
-    long seconds = millis / NUM_MILLIS_PER_SECOND;
-    int nanos = (int) (millis % NUM_MILLIS_PER_SECOND) * NUM_NANOS_PER_MILLI;
-    if (nanos < 0) {
-      return new Timestamp(seconds - 1, (int) (nanos + NUM_NANOS_PER_SECOND));
-    } else {
-      return new Timestamp(seconds, nanos);
-    }
+  public static Timestamp fromMillis(long epochMilli) {
+    long secs = floorDiv(epochMilli, MILLIS_PER_SECOND);
+    int mos = (int) floorMod(epochMilli, MILLIS_PER_SECOND);
```
Comment (sebright) [inline api/src/main/java/io/opencensus/common/Timestamp.java]: Should this variable be "mod"?

File: api/src/main/java/io/opencensus/common/Timestamp.java [hunk 2]
```diff
@@ -45,101 +46,116 @@ private Timestamp(long seconds, int nanos) {
...
-  public boolean equals(Object obj) {
-    if (obj == this) {
-      return true;
+  /**
+   * Returns a {@link Duration} calculated as: {@code this - timestamp}.
+   *
+   * @param timestamp the {@code Timestamp} to subtract.
+   * @return the calculated {@code Duration}. For invalid inputs, {@link Duration#ZERO} is returned.
+   */
+  public Duration minusTimestamp(Timestamp timestamp) {
```
Comment (sebright) [inline api/src/main/java/io/opencensus/common/Timestamp.java]: How about subtractTimestamp?  I think it is more consistent with "add" in addDuration.

File: api/src/main/java/io/opencensus/common/Duration.java [hunk 3]
```diff
@@ -29,88 +41,48 @@
    *     negative `nanos` field. For durations of one second or more, a non-zero value for the
    *     `nanos` field must be of the same sign as the `seconds` field. Must be from -999,999,999 to
    *     +999,999,999 inclusive.
-   * @return new {@link Duration} with specified fields. For invalid inputs, a {@link Duration} of
-   *     zero is returned.
+   * @return new {@code Duration} with specified fields. For invalid inputs, {@link #ZERO} is
+   *     returned.
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/common/Duration.java]: Done.

File: api/src/main/java/io/opencensus/common/Timestamp.java (hunk 1 above)
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/common/Timestamp.java]: millis of second

File: api/src/main/java/io/opencensus/common/Timestamp.java (hunk 2 above)
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/common/Timestamp.java]: Done.

**Comment by bogdandrutu** [comment]
PTAL.

File: api/src/main/java/io/opencensus/common/Duration.java [hunk 4]
```diff
@@ -13,12 +13,24 @@
...
 /**
  * Represents a signed, fixed-length span of time represented as a count of seconds and fractions of
  * seconds at nanosecond resolution. It is independent of any calendar and concepts like "day" or
  * "month". Range is approximately +-10,000 years.
  */
-public class Duration {
+@Immutable
+@AutoValue
+public abstract class Duration {
+  public static final Duration ZERO = new AutoValue_Duration(0, 0);
```
Comment (sebright) [inline api/src/main/java/io/opencensus/common/Duration.java]: Should this still be public?

Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/common/Duration.java]: Done.

**Pull request merged**